File: runtime/go/prompty/model/memory.go
package prompty

import (
	"fmt"
	"strings"
	"unicode"
)

// ScoredMemory is a recalled memory paired with its deterministic lexical score.
type ScoredMemory struct {
	Entry          MemoryEntry
	Score          float64
	KeywordMatches int
}

// MemoryPort is host-owned persistence for a whole MemoryStore snapshot.
type MemoryPort interface {
	Load() MemoryStore
	Save(store MemoryStore)
}

// AddMemory appends a memory with no tier policy applied.
func AddMemory(store *MemoryStore, entry MemoryEntry) {
	store.Entries = append(store.Entries, entry)
}

// Remember inserts a memory applying core deduplication and cap eviction.
func Remember(store *MemoryStore, entry MemoryEntry, maxEntries int) {
	if entry.Category == MemoryCategoryCore {
		filtered := store.Entries[:0]
		for _, existing := range store.Entries {
			if existing.Category == MemoryCategoryCore && tagsEqual(existing.Tags, entry.Tags) {
				continue
			}
			filtered = append(filtered, existing)
		}
		store.Entries = filtered
	}
	store.Entries = append(store.Entries, entry)
	EvictToCap(store, maxEntries)
}

// EvictToCap evicts oldest archival memories first, then oldest remaining memories.
func EvictToCap(store *MemoryStore, maxEntries int) int {
	if maxEntries <= 0 {
		return 0
	}
	evicted := 0
	for len(store.Entries) > maxEntries {
		victim := 0
		for i, entry := range store.Entries {
			if entry.Category == MemoryCategoryArchival {
				victim = i
				break
			}
		}
		store.Entries = append(store.Entries[:victim], store.Entries[victim+1:]...)
		evicted++
	}
	return evicted
}

// UpdateMemory replaces the memory at index.
func UpdateMemory(store *MemoryStore, index int, entry MemoryEntry) error {
	if index < 0 || index >= len(store.Entries) {
		return fmt.Errorf("memory index %d out of bounds (len %d)", index, len(store.Entries))
	}
	store.Entries[index] = entry
	return nil
}

// UpdateContent replaces only the content at index.
func UpdateContent(store *MemoryStore, index int, content string) error {
	if index < 0 || index >= len(store.Entries) {
		return fmt.Errorf("memory index %d out of bounds (len %d)", index, len(store.Entries))
	}
	store.Entries[index].Content = content
	return nil
}

// RemoveMemory removes and returns the memory at index.
func RemoveMemory(store *MemoryStore, index int) (MemoryEntry, error) {
	if index < 0 || index >= len(store.Entries) {
		return MemoryEntry{}, fmt.Errorf("memory index %d out of bounds (len %d)", index, len(store.Entries))
	}
	entry := store.Entries[index]
	store.Entries = append(store.Entries[:index], store.Entries[index+1:]...)
	return entry, nil
}

// ClearMemory removes memories by category, or all memories when category is nil.
func ClearMemory(store *MemoryStore, category *MemoryCategory) int {
	before := len(store.Entries)
	if category == nil {
		store.Entries = nil
		return before
	}
	filtered := store.Entries[:0]
	for _, entry := range store.Entries {
		if entry.Category != *category {
			filtered = append(filtered, entry)
		}
	}
	store.Entries = filtered
	return before - len(store.Entries)
}

// CoreMemories returns core memories in insertion order.
func CoreMemories(store MemoryStore) []MemoryEntry {
	var core []MemoryEntry
	for _, entry := range store.Entries {
		if entry.Category == MemoryCategoryCore {
			core = append(core, entry)
		}
	}
	return core
}

// Recall returns deterministic lexical recall results.
func Recall(store MemoryStore, query string, limit int) []ScoredMemory {
	tokens := queryTokens(query)
	hasQuery := len(tokens) > 0
	scored := make([]ScoredMemory, 0, len(store.Entries))
	for _, entry := range store.Entries {
		score, matches := scoreEntry(entry, tokens)
		if hasQuery && matches == 0 {
			continue
		}
		scored = append(scored, ScoredMemory{Entry: entry, Score: score, KeywordMatches: matches})
	}
	stableSortByScore(scored)
	if limit > 0 && len(scored) > limit {
		return scored[:limit]
	}
	return scored
}

// FormatForSystemPrompt formats core memories for injection into a system prompt.
func FormatForSystemPrompt(store MemoryStore) string {
	core := CoreMemories(store)
	if len(core) == 0 {
		return ""
	}
	var out strings.Builder
	out.WriteString("## Memory\n")
	for _, entry := range core {
		out.WriteString("- ")
		out.WriteString(entry.Content)
		out.WriteByte('\n')
	}
	return out.String()
}

// FormatRecallResults formats a recall result set for presentation.
func FormatRecallResults(results []ScoredMemory) string {
	if len(results) == 0 {
		return ""
	}
	var out strings.Builder
	for i, result := range results {
		out.WriteString(fmt.Sprintf("%d. [%s] %s\n", i+1, result.Entry.Category, result.Entry.Content))
		if len(result.Entry.Tags) > 0 {
			out.WriteString("   tags: ")
			out.WriteString(strings.Join(result.Entry.Tags, ", "))
			out.WriteByte('\n')
		}
	}
	return out.String()
}

func tagsEqual(left []string, right []string) bool {
	if len(left) != len(right) {
		return false
	}
	for i := range left {
		if left[i] != right[i] {
			return false
		}
	}
	return true
}

func queryTokens(query string) []string {
	tokens := []string{}
	seen := map[string]bool{}
	for _, raw := range strings.Fields(query) {
		token := strings.ToLower(trimNonAlphanumeric(raw))
		if token != "" && !seen[token] {
			seen[token] = true
			tokens = append(tokens, token)
		}
	}
	return tokens
}

func trimNonAlphanumeric(value string) string {
	runes := []rune(value)
	start := 0
	end := len(runes)
	for start < end && !unicode.IsLetter(runes[start]) && !unicode.IsDigit(runes[start]) {
		start++
	}
	for end > start && !unicode.IsLetter(runes[end-1]) && !unicode.IsDigit(runes[end-1]) {
		end--
	}
	return string(runes[start:end])
}

func scoreEntry(entry MemoryEntry, tokens []string) (float64, int) {
	if len(tokens) == 0 {
		return 0, 0
	}
	content := strings.ToLower(entry.Content)
	tags := make([]string, len(entry.Tags))
	for i, tag := range entry.Tags {
		tags[i] = strings.ToLower(tag)
	}
	score := 0.0
	matches := 0
	for _, token := range tokens {
		inContent := strings.Contains(content, token)
		inTags := false
		for _, tag := range tags {
			if strings.Contains(tag, token) {
				inTags = true
				break
			}
		}
		if inContent || inTags {
			matches++
		}
		if inContent {
			score += 2
		}
		if inTags {
			score += 3
		}
	}
	if score > 0 && entry.Category == MemoryCategoryCore {
		score += 1
	}
	return score, matches
}

func stableSortByScore(values []ScoredMemory) {
	for i := 1; i < len(values); i++ {
		item := values[i]
		j := i - 1
		for j >= 0 && values[j].Score < item.Score {
			values[j+1] = values[j]
			j--
		}
		values[j+1] = item
	}
}

File: runtime/csharp/Prompty.Core/Memory.cs
#nullable enable

using System.Text;
using System.Linq;

namespace Prompty.Core;

public sealed record ScoredMemory(MemoryEntry Entry, double Score, int KeywordMatches);

public interface IMemoryPort
{
    MemoryStore Load();

    void Save(MemoryStore store);
}

public static class Memory
{
    public static void Add(MemoryStore store, MemoryEntry entry) => store.Entries.Add(entry);

    public static void Remember(MemoryStore store, MemoryEntry entry, int maxEntries = 0)
    {
        if (entry.Category == MemoryCategory.Core)
        {
            store.Entries = store.Entries
                .Where(existing => existing.Category != MemoryCategory.Core || !TagsEqual(existing.Tags, entry.Tags))
                .ToList();
        }

        store.Entries.Add(entry);
        EvictToCap(store, maxEntries);
    }

    public static int EvictToCap(MemoryStore store, int maxEntries)
    {
        if (maxEntries <= 0)
            return 0;

        var evicted = 0;
        while (store.Entries.Count > maxEntries)
        {
            var victim = 0;
            for (var i = 0; i < store.Entries.Count; i++)
            {
                if (store.Entries[i].Category == MemoryCategory.Archival)
                {
                    victim = i;
                    break;
                }
            }

            store.Entries.RemoveAt(victim);
            evicted++;
        }

        return evicted;
    }

    public static void Update(MemoryStore store, int index, MemoryEntry entry)
    {
        RequireIndex(store, index);
        store.Entries[index] = entry;
    }

    public static void UpdateContent(MemoryStore store, int index, string content)
    {
        RequireIndex(store, index);
        store.Entries[index].Content = content;
    }

    public static MemoryEntry Remove(MemoryStore store, int index)
    {
        RequireIndex(store, index);
        var entry = store.Entries[index];
        store.Entries.RemoveAt(index);
        return entry;
    }

    public static int Clear(MemoryStore store, MemoryCategory? category = null)
    {
        var before = store.Entries.Count;
        if (category is null)
        {
            store.Entries.Clear();
        }
        else
        {
            store.Entries = store.Entries.Where(entry => entry.Category != category).ToList();
        }

        return before - store.Entries.Count;
    }

    public static IReadOnlyList<MemoryEntry> CoreMemories(MemoryStore store) =>
        store.Entries.Where(entry => entry.Category == MemoryCategory.Core).ToList();

    public static IReadOnlyList<ScoredMemory> Recall(MemoryStore store, string? query, int limit = 0)
    {
        var tokens = QueryTokens(query);
        var hasQuery = tokens.Count > 0;
        var scored = new List<ScoredMemory>();

        foreach (var entry in store.Entries)
        {
            var (score, matches) = ScoreEntry(entry, tokens);
            if (hasQuery && matches == 0)
                continue;
            scored.Add(new ScoredMemory(entry, score, matches));
        }

        scored = scored.OrderByDescending(item => item.Score).ToList();
        return limit > 0 && scored.Count > limit ? scored.Take(limit).ToList() : scored;
    }

    public static string FormatForSystemPrompt(MemoryStore store)
    {
        var core = CoreMemories(store);
        if (core.Count == 0)
            return string.Empty;

        var output = new StringBuilder("## Memory\n");
        foreach (var entry in core)
        {
            output.Append("- ").Append(entry.Content).Append('\n');
        }

        return output.ToString();
    }

    public static string FormatRecallResults(IEnumerable<ScoredMemory> results)
    {
        var output = new StringBuilder();
        var index = 0;
        foreach (var result in results)
        {
            index++;
            output
                .Append(index)
                .Append(". [")
                .Append(MemoryCategoryParser.ToValue(result.Entry.Category))
                .Append("] ")
                .Append(result.Entry.Content)
                .Append('\n');

            if (result.Entry.Tags is { Count: > 0 })
            {
                output.Append("   tags: ").Append(string.Join(", ", result.Entry.Tags)).Append('\n');
            }
        }

        return output.ToString();
    }

    private static void RequireIndex(MemoryStore store, int index)
    {
        if (index < 0 || index >= store.Entries.Count)
            throw new ArgumentOutOfRangeException(nameof(index), $"memory index {index} out of bounds (len {store.Entries.Count})");
    }

    private static bool TagsEqual(IList<string>? left, IList<string>? right)
    {
        var a = left ?? [];
        var b = right ?? [];
        return a.SequenceEqual(b);
    }

    private static List<string> QueryTokens(string? query)
    {
        var tokens = new List<string>();
        foreach (var token in (query ?? string.Empty)
            .Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries)
            .Select(raw => TrimNonAlphanumeric(raw).ToLowerInvariant())
            .Where(token => token.Length > 0))
        {
            if (!tokens.Contains(token))
            {
                tokens.Add(token);
            }
        }

        return tokens;
    }

    private static string TrimNonAlphanumeric(string value)
    {
        var start = 0;
        var end = value.Length;
        while (start < end && !char.IsLetterOrDigit(value[start]))
            start++;
        while (end > start && !char.IsLetterOrDigit(value[end - 1]))
            end--;
        return value[start..end];
    }

    private static (double Score, int Matches) ScoreEntry(MemoryEntry entry, IReadOnlyList<string> tokens)
    {
        if (tokens.Count == 0)
            return (0, 0);

        var content = (entry.Content ?? string.Empty).ToLowerInvariant();
        var tags = (entry.Tags ?? []).Select(tag => tag.ToLowerInvariant()).ToList();
        var score = 0.0;
        var matches = 0;

        foreach (var token in tokens)
        {
            var inContent = content.Contains(token, StringComparison.Ordinal);
            var inTags = tags.Any(tag => tag.Contains(token, StringComparison.Ordinal));
            if (inContent || inTags)
                matches++;
            if (inContent)
                score += 2;
            if (inTags)
                score += 3;
        }

        if (score > 0 && entry.Category == MemoryCategory.Core)
            score += 1;

        return (score, matches);
    }
}
